src/common/Singleton.h

```diff
@@ -26,20 +28,24 @@ class Singleton {
   Singleton() noexcept = default;
 
   Singleton(const Singleton&) = delete;
-
   Singleton& operator=(const Singleton&) = delete;
 
+  Singleton(Singleton&&) = delete;
+  Singleton& operator=(Singleton&&) = delete;
+
   virtual ~Singleton() = default;  // to silence base class Singleton<T> has a
                                    // non-virtual destructor [-Weffc++]
 
  public:
-  static T& GetInstance() noexcept(std::is_nothrow_constructible<T>::value) {
-    // Guaranteed to be destroyed.
-    // Instantiated on first use.
-    // Thread safe in C++11
-    static T instance;
-
-    return instance;
+  static T& GetInstance(const std::function<std::shared_ptr<T>()>& _allocator =
+                            []() { return std::make_shared<T>(); }) {
```

Comment:
Do you plan to use the `allocator`?  I quite strongly feel about the YAGNI principle and think things should be added only when they are needed, not proactively.

Reply:
Overriding the default allocater in the singleton is kind of a hack, but gave me to possibility to inject a different class to override the CallRunner method. 
Would be better to remove all singletons and use a different design that will give use the override possibility to test and implement different versions.